File: scripts/analyze_duplicates.py
import json
import re
import sys

# Path to the data file
DATA_FILE = "/Users/am/.gemini/antigravity/brain/3f6fd10b-6b20-4ac0-9534-c253fa72e406/.system_generated/steps/36/output.txt"

def load_data(file_path):
    try:
        with open(file_path, 'r') as f:
            content = f.read()
    except FileNotFoundError:
        print(f"File not found: {file_path}")
        return []

    # Try to find the JSON array list
    match = re.search(r'\[.*\]', content, re.DOTALL)
    if not match:
        print("Could not find JSON data ([...]) in file.")
        return []
    
    json_candidates = []
    raw_match = match.group(0)
    json_candidates.append(raw_match)
    
    # Candidate 2: Unescape quotes if they look escaped
    if '\\"' in raw_match:
        unescaped = raw_match.replace('\\"', '"')
        # We might also need to handle other escapes if necessary, but start with quotes
        json_candidates.append(unescaped)

    data = None
    for i, candidate in enumerate(json_candidates):
        try:
            data = json.loads(candidate)
            print(f"Successfully loaded data using candidate {i+1}")
            break
        except json.JSONDecodeError as e:
            print(f"Candidate {i+1} failed JSON decode: {e}")
            # print(f"Snippet: {candidate[:100]}...")

    if not data:
        print("Failed to decode JSON from any candidate.")
        print(f"Raw snippet: {raw_match[:200]}")
        return []

    return data

def normalize_string(s):
    if not s:
        return ""
    return str(s).lower().strip() # Ensure it's string (e.g. if numbers)

def calculate_completeness_score(record):
    score = 0
    # Weighted fields
    # document: strict check for non-empty string that isn't just whitespace
    doc = record.get('documento')
    if doc and str(doc).strip(): score += 5
    
    email = record.get('email')
    if email and str(email).strip(): score += 3
    
    phone = record.get('telefono')
    if phone and str(phone).strip(): score += 3
    
    if record.get('fecha_nacimiento'): score += 2
    if record.get('link_historia_clinica'): score += 2
    
    if record.get('presupuesto_total') and float(record['presupuesto_total'] or 0) > 0: score += 1
    if record.get('financ_monto_total') and float(record['financ_monto_total'] or 0) > 0: score += 1
    
    # Prefer newer records if scores are tied? Or older?
    # Usually older records have more history, but newer might be cleaner.
    # Let's add a tiny fraction based on ID or date to break ties consistently
    return score

def main():
    data = load_data(DATA_FILE)
    if not data:
        return

    print(f"Loaded {len(data)} records.")

    # Group by normalized name + surname
    groups = {}
    for record in data:
        name = normalize_string(record.get('nombre'))
        surname = normalize_string(record.get('apellido'))
        
        # Skip if name/surname is extremely short or empty
        if len(name) < 2 or len(surname) < 2:
            continue

        key = f"{name}|{surname}"
        
        # Also, check if 'documento' is same?
        # A person might be entered as "Juan Perez" and "Juan Antonio Perez" - this is harder.
        # But let's stick to exact name+surname match for now as requested by user logic typically
        
        if key not in groups:
            groups[key] = []
        groups[key].append(record)

    duplicate_groups = {k: v for k, v in groups.items() if len(v) > 1}
    
    print(f"Found {len(duplicate_groups)} groups with duplicates based on Name + Surname.")

    to_delete = []
    to_keep = []
    
    summary_counts = {"kept": 0, "deleted": 0}

    # Open a report file
    with open('duplicate_report.txt', 'w') as report:
        report.write("--- Duplicate Report ---\n")
        
        for key, group in duplicate_groups.items():
            # Score each record
            scored_group = []
            for record in group:
                score = calculate_completeness_score(record)
                scored_group.append((score, record))
            
            # Sort by score descending. 
            # If tie, use ID or created_at to be deterministic?
            # Let's use ID as secondary sort key
            scored_group.sort(key=lambda x: (x[0], x[1].get('id_paciente')), reverse=True)
            
            winner = scored_group[0][1]
            losers = [x[1] for x in scored_group[1:]]
            
            to_keep.append(winner)
            to_delete.extend(losers)
            summary_counts["kept"] += 1
            summary_counts["deleted"] += len(losers)
    
            report.write(f"\nGroup: {key.replace('|', ' ')}\n")
            report.write(f"  KEEP: ID={winner['id_paciente']} (Score: {scored_group[0][0]}) - Doc: {winner.get('documento')}, Email: {winner.get('email')}, Updated: {winner.get('updated_at')}\n")
            for loser in losers:
                l_score = calculate_completeness_score(loser)
                report.write(f"  DELETE: ID={loser['id_paciente']} (Score: {l_score}) - Doc: {loser.get('documento')}, Email: {loser.get('email')}, Updated: {loser.get('updated_at')}\n")

    print(f"Total duplicates found: {summary_counts['deleted']}")
    print("Report generated in 'duplicate_report.txt'")

    # Generate SQL delete statements
    if to_delete:
        with open('delete_duplicates.sql', 'w') as sql_file:
            ids_to_delete = [f"'{r['id_paciente']}'" for r in to_delete]
            batch_size = 50
            for i in range(0, len(ids_to_delete), batch_size):
                batch = ids_to_delete[i:i+batch_size]
                sql_file.write(f"UPDATE pacientes SET is_deleted = true WHERE id_paciente IN ({', '.join(batch)});\n")
        print("SQL script generated in 'delete_duplicates.sql'")

if __name__ == "__main__":
    main()

File: scripts/analyze_duplicates.ts

import { createClient } from '@supabase/supabase-js';
import * as fs from 'fs';
import * as path from 'path';
import * as dotenv from 'dotenv';

// Load .env.local first, then .env
dotenv.config({ path: path.join(process.cwd(), '.env.local') });
dotenv.config({ path: path.join(process.cwd(), '.env') });


// Initialize Supabase
const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL;
const supabaseServiceKey = process.env.SUPABASE_SERVICE_ROLE_KEY; // Use service role for admin tasks

if (!supabaseUrl || !supabaseServiceKey) {
    console.error('Missing SUPABASE_URL or SUPABASE_SERVICE_ROLE_KEY');
    process.exit(1);
}

const supabase = createClient(supabaseUrl, supabaseServiceKey);

/**
 * Validates if a string is a valid UUID.
 */
function isValidUuid(id: string): boolean {
    const uuidRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;
    return uuidRegex.test(id);
}

async function main() {
    console.log('═══════════════════════════════════════════════════════════');
    console.log('PASO 1: BACKUP DE SEGURIDAD');
    console.log('───────────────────────────');

    // 1. Fetch all patients
    console.log('Fetching all patients...');
    let allPatients: any[] = [];
    let from = 0;
    const pageSize = 1000;
    let hasMore = true;

    while (hasMore) {
        const { data, error } = await supabase
            .from('pacientes')
            .select('*')
            .eq('is_deleted', false)
            .range(from, from + pageSize - 1);

        if (error) {
            console.error('Error fetching patients:', error);
            process.exit(1);
        }

        if (data) {
            allPatients = [...allPatients, ...data];
            if (data.length < pageSize) {
                hasMore = false;
            } else {
                from += pageSize;
            }
        } else {
            hasMore = false;
        }
    }

    const patients = allPatients;

    if (!patients || patients.length === 0) {
        console.log('No patients found.');
        process.exit(0);
    }

    console.log(`Found ${patients.length} patients.`);

    // 2. Backup
    const dateStr = new Date().toISOString().split('T')[0];
    const backupFileName = `backup_pacientes_${dateStr}.json`;
    const backupPath = path.join(process.cwd(), backupFileName);

    fs.writeFileSync(backupPath, JSON.stringify(patients, null, 2));
    console.log(`Backup saved to: ${backupPath}`);
    console.log('Backup completo.');
    console.log('');

    console.log('═══════════════════════════════════════════════════════════');
    console.log('PASO 2: ANÁLISIS DE DUPLICADOS');
    console.log('───────────────────────────────');

    // Helper to normalize strings for comparison
    const normalize = (s: string | null) => s ? s.toLowerCase().trim() : '';
    const normalizePhone = (s: string | null) => s ? s.replace(/\D/g, '') : '';

    // Data structures for grouping
    const groups: Record<string, any[]> = {};

    // Grouping
    patients.forEach(p => {
        const normName = normalize(p.nombre);
        const normSurname = normalize(p.apellido);

        if (!normName || !normSurname) return;

        // Primary key: Normalized Name + Surname
        // We group by Name+Surname first. This is safer now because 
        // we have massive duplicates with identical names.
        const key = `${normName}|${normSurname}`;

        if (!groups[key]) groups[key] = [];
        groups[key].push(p);
    });

    // Sub-segmenting or combining groups based on DNI/Email if necessary?
    // Actually, users' instruction "Nombre completo + Teléfono" suggests 
    // that if names match but phones are different, they might be different.
    // However, if we have 151 records with same name, they are duplicates.

    const finalGroups = Object.values(groups).filter(g => g.length > 1);



    const duplicateGroups: any[][] = finalGroups;

    console.log(`Found ${duplicateGroups.length} groups of duplicates.`);
    console.log('');



    console.log('═══════════════════════════════════════════════════════════');
    console.log('PASO 3 & 4: DETERMINAR QUÉ CONSERVAR Y REPORTE');
    console.log('──────────────────────────────────────────────');

    // Relational data counts
    const fetchCounts = async (table: string, idField: string) => {
        let allData: any[] = [];
        let from = 0;
        const pageSize = 1000;
        let hasMore = true;

        while (hasMore) {
            const { data, error } = await supabase.from(table).select(idField).range(from, from + pageSize - 1);
            if (error) break;
            if (data) {
                allData = [...allData, ...data];
                if (data.length < pageSize) hasMore = false;
                else from += pageSize;
            } else hasMore = false;
        }

        const counts: Record<string, number> = {};
        allData.forEach((r: any) => {
            const id = r[idField];
            counts[id] = (counts[id] || 0) + 1;
        });
        return counts;
    };

    console.log('Checking related data (movements, history, plans)...');
    const movementCounts = await fetchCounts('caja_recepcion_movimientos', 'paciente_id');
    const clinicalCounts = await fetchCounts('historia_clinica', 'paciente_id');
    const planCounts = await fetchCounts('planes_tratamiento', 'paciente_id');

    interface ReportRow {
        DNI: string;
        Nombre: string;
        'Duplicados Encontrados': string;
        'Conservar (ID)': string;
        'Eliminar (IDs)': string;
    }

    const reportData: ReportRow[] = [];
    const allLoserIds: string[] = [];

    for (const group of duplicateGroups) {
        // 1. Calculate Score
        const scored = group.map(p => {
            const id = p.id_paciente;
            let score = 0;

            // Fields score
            Object.keys(p).forEach(k => {
                if (p[k] !== null && p[k] !== '' && p[k] !== undefined) score++;
            });

            // Activity score (Value historical data highly)
            const hasMovements = (movementCounts[id] || 0) > 0;
            const hasClinical = (clinicalCounts[id] || 0) > 0;
            const hasPlans = (planCounts[id] || 0) > 0;

            // Weight activity heavily - if they have clinical history or payments, they are likely the "real" one
            if (hasMovements) score += 100;
            if (hasClinical) score += 200;
            if (hasPlans) score += 150;

            return { ...p, _score: score, _hasActivity: hasMovements || hasClinical || hasPlans };
        });

        // 2. Sort
        scored.sort((a, b) => {
            // 1. Activity + Completeness
            if (b._score !== a._score) return b._score - a._score;

            // 2. Recency (Tie-breaker)
            const dateA = new Date(a.updated_at || a.created_at || a.fecha_alta || 0).getTime();
            const dateB = new Date(b.updated_at || b.created_at || b.fecha_alta || 0).getTime();
            return dateB - dateA;
        });

        const winner = scored[0];
        const losers = scored.slice(1);

        allLoserIds.push(...losers.map(l => l.id_paciente));

        reportData.push({
            DNI: winner.documento || '(Sin DNI)',
            Nombre: `${winner.nombre} ${winner.apellido}`,
            'Duplicados Encontrados': `${group.length} registros`,
            'Conservar (ID)': winner.id_paciente,
            'Eliminar (IDs)': losers.map(l => l.id_paciente.substring(0, 8)).join(', ') // Short IDs for table display
        });
    }

    // Final Summary
    console.log(`\nTotal de registros a eliminar: ${allLoserIds.length}\n`);

    // Show First 20 as Table for Preview
    const preview = reportData.slice(0, 20);
    console.table(preview);

    if (reportData.length > 20) {
        console.log(`... y ${reportData.length - 20} grupos más.`);
    }

    // Save detailed migration script (SQL)
    const sql = `-- MIGRACION: ELIMINACION DE DUPLICADOS PACIENTES
-- Generado: ${new Date().toLocaleString()}
-- Total a eliminar: ${allLoserIds.length}

BEGIN;

-- Marcar como eliminados (Soft delete preferible primero para seguridad)
UPDATE pacientes 
SET is_deleted = true, 
    deleted_at = now(), 
    delete_reason = 'Limpieza de duplicados masiva'
WHERE id_paciente IN (
  ${allLoserIds.filter(isValidUuid).map(id => `'${id}'`).join(',\n  ')}
);

-- O eliminacion fisica si el usuario confirma:
-- DELETE FROM pacientes WHERE id_paciente IN (...);

COMMIT;
`;

    const sqlPath = path.join(process.cwd(), `delete_duplicates_${dateStr}.sql`);
    fs.writeFileSync(sqlPath, sql);
    console.log(`\nScript SQL de eliminación generado en: ${sqlPath}`);

    const reportPath = path.join(process.cwd(), `reporte_duplicados_${dateStr}.csv`);
    const csvContent = "DNI,Nombre,Duplicados,Conservar_ID,Eliminar_IDs\n" +
        reportData.map(r => `${r.DNI},"${r.Nombre}",${r['Duplicados Encontrados']},${r['Conservar (ID)']},"${r['Eliminar (IDs)']}"`).join('\n');
    fs.writeFileSync(reportPath, csvContent);
    console.log(`Reporte CSV completo guardado en: ${reportPath}`);
}

main().catch(console.error);
